File: macos-helper/Tests/CaptureCoreTests/SpoolStoreTests.swift
import Foundation
import Testing
@testable import CaptureCore

@Suite final class SpoolStoreTests {

    private let rootDir: URL
    private let store: SpoolStore

    init() throws {
        rootDir = FileManager.default.temporaryDirectory
            .appendingPathComponent("agb-store-tests-\(UUID().uuidString)/spool", isDirectory: true)
        store = try SpoolStore(rootURL: rootDir)
    }

    deinit {
        try? FileManager.default.removeItem(at: rootDir.deletingLastPathComponent())
    }

    @Test func createListDelete() throws {
        #expect(try store.sessionDirectories().isEmpty)

        let spoolerA = try store.createSession(localId: "aaa", startedAt: Date(), sourceApp: nil)
        let spoolerB = try store.createSession(localId: "bbb", startedAt: Date(), sourceApp: "Zoom")
        #expect(try store.sessionDirectories().count == 2)
        #expect(FileManager.default.fileExists(
            atPath: spoolerA.directory.appendingPathComponent("manifest.json").path))

        try store.deleteSession(spoolerA)
        let remaining = try store.sessionDirectories()
        #expect(remaining.count == 1)
        #expect(remaining.first?.lastPathComponent == "session-bbb")
        #expect(spoolerB.snapshot.sourceApp == "Zoom")
    }

    @Test func sessionDirectoryPermissionsAre0700() throws {
        let spooler = try store.createSession(localId: "perm", startedAt: Date(), sourceApp: nil)
        let attrs = try FileManager.default.attributesOfItem(atPath: spooler.directory.path)
        #expect((attrs[.posixPermissions] as? NSNumber)?.intValue == 0o700)
    }

    @Test func pendingDetectionExcludesFinalizedAndSortsOldestFirst() throws {
        _ = try store.createSession(localId: "old",
                                    startedAt: Date(timeIntervalSinceNow: -600),
                                    sourceApp: nil)
        _ = try store.createSession(localId: "recent",
                                    startedAt: Date(timeIntervalSinceNow: -60),
                                    sourceApp: nil)
        let done = try store.createSession(localId: "done",
                                           startedAt: Date(timeIntervalSinceNow: -30),
                                           sourceApp: nil)
        try done.markFinalized()

        let pending = try store.pendingSessions()
        #expect(pending.map(\.localId) == ["old", "recent"], "oldest first, finalized excluded")
    }

    @Test func openSessionReturnsSharedCachedInstance() throws {
        let created = try store.createSession(localId: "shared", startedAt: Date(), sourceApp: nil)
        let opened = try store.openSession(at: created.directory)
        #expect(created === opened, "writer and uploader must share one spooler instance")
    }

    @Test func openSessionFromDiskAfterRestart() throws {
        let created = try store.createSession(localId: "restart", startedAt: Date(),
                                              sourceApp: nil, chunkSeconds: 1)
        try created.append(Data(count: AudioConstants.bytesPerSecond))
        let dir = created.directory

        // New store = new process.
        let freshStore = try SpoolStore(rootURL: rootDir)
        let reopened = try freshStore.openSession(at: dir)
        #expect(!(created === reopened))
        #expect(reopened.snapshot.seqsWritten == [0])
    }

    @Test func adoptOrphansMarksEndedAndPartial() throws {
        let orphan = try store.createSession(localId: "orphan", startedAt: Date(),
                                             sourceApp: nil, chunkSeconds: 1)
        try orphan.append(Data(count: AudioConstants.bytesPerSecond * 2))
        #expect(orphan.snapshot.endedAt == nil)

        let active = try store.createSession(localId: "active", startedAt: Date(), sourceApp: nil)

        let adopted = try store.adoptOrphans(activeLocalIds: ["active"])
        #expect(adopted.map(\.localId) == ["orphan"])

        let snap = orphan.snapshot
        #expect(snap.endedAt != nil)
        #expect(snap.partial)
        #expect(snap.durationSecs == 2)
        #expect(active.snapshot.endedAt == nil, "active session must not be adopted")
    }

    @Test func corruptManifestIsSkippedNotFatal() throws {
        _ = try store.createSession(localId: "good", startedAt: Date(), sourceApp: nil)
        let badDir = rootDir.appendingPathComponent("session-corrupt", isDirectory: true)
        try FileManager.default.createDirectory(at: badDir, withIntermediateDirectories: true)
        try Data("not json".utf8).write(to: badDir.appendingPathComponent("manifest.json"))

        let pending = try store.pendingSessions()
        #expect(pending.map(\.localId) == ["good"])
    }
}

File: windows-helper/CaptureCore.Tests/SpoolStoreTests.cs
using AGB.CaptureCore;
using Xunit;

namespace AGB.CaptureCore.Tests;

/// <summary>Port of <c>SpoolStoreTests.swift</c> — create/list/delete, shared instances, orphan adoption.</summary>
public class SpoolStoreTests : IDisposable
{
    private readonly string _rootDir;
    private readonly SpoolStore _store;

    public SpoolStoreTests()
    {
        _rootDir = Path.Combine(Path.GetTempPath(), $"agb-store-tests-{Guid.NewGuid()}", "spool");
        _store = new SpoolStore(_rootDir);
    }

    public void Dispose()
    {
        try { Directory.Delete(Path.GetDirectoryName(_rootDir)!, recursive: true); } catch { /* ignore */ }
    }

    [Fact]
    public void CreateListDelete()
    {
        Assert.Empty(_store.SessionDirectories());

        var spoolerA = _store.CreateSession(localId: "aaa", startedAt: DateTimeOffset.UtcNow, sourceApp: null);
        var spoolerB = _store.CreateSession(localId: "bbb", startedAt: DateTimeOffset.UtcNow, sourceApp: "Zoom");
        Assert.Equal(2, _store.SessionDirectories().Count);
        Assert.True(File.Exists(Path.Combine(spoolerA.Directory, "manifest.json")));

        _store.DeleteSession(spoolerA);
        var remaining = _store.SessionDirectories();
        Assert.Single(remaining);
        Assert.Equal("session-bbb", Path.GetFileName(remaining[0]));
        Assert.Equal("Zoom", spoolerB.Snapshot.SourceApp);
    }

    [Fact]
    public void PendingDetectionExcludesFinalizedAndSortsOldestFirst()
    {
        _store.CreateSession(localId: "old", startedAt: DateTimeOffset.UtcNow.AddSeconds(-600), sourceApp: null);
        _store.CreateSession(localId: "recent", startedAt: DateTimeOffset.UtcNow.AddSeconds(-60), sourceApp: null);
        var done = _store.CreateSession(localId: "done", startedAt: DateTimeOffset.UtcNow.AddSeconds(-30), sourceApp: null);
        done.MarkFinalized();

        var pending = _store.PendingSessions();
        Assert.Equal(new[] { "old", "recent" }, pending.Select(s => s.LocalId)); // oldest first, finalized excluded
    }

    [Fact]
    public void OpenSessionReturnsSharedCachedInstance()
    {
        var created = _store.CreateSession(localId: "shared", startedAt: DateTimeOffset.UtcNow, sourceApp: null);
        var opened = _store.OpenSession(created.Directory);
        Assert.Same(created, opened); // writer and uploader must share one spooler instance
    }

    [Fact]
    public void OpenSessionFromDiskAfterRestart()
    {
        var created = _store.CreateSession(localId: "restart", startedAt: DateTimeOffset.UtcNow, sourceApp: null, chunkSeconds: 1);
        created.Append(new byte[AudioConstants.BytesPerSecond]);
        string dir = created.Directory;

        // New store = new process.
        var freshStore = new SpoolStore(_rootDir);
        var reopened = freshStore.OpenSession(dir);
        Assert.NotSame(created, reopened);
        Assert.Equal(new[] { 0 }, reopened.Snapshot.SeqsWritten);
    }

    [Fact]
    public void AdoptOrphansMarksEndedAndPartial()
    {
        var orphan = _store.CreateSession(localId: "orphan", startedAt: DateTimeOffset.UtcNow, sourceApp: null, chunkSeconds: 1);
        orphan.Append(new byte[AudioConstants.BytesPerSecond * 2]);
        Assert.Null(orphan.Snapshot.EndedAt);

        var active = _store.CreateSession(localId: "active", startedAt: DateTimeOffset.UtcNow, sourceApp: null);

        var adopted = _store.AdoptOrphans(new HashSet<string> { "active" });
        Assert.Equal(new[] { "orphan" }, adopted.Select(s => s.LocalId));

        var snap = orphan.Snapshot;
        Assert.NotNull(snap.EndedAt);
        Assert.True(snap.Partial);
        Assert.Equal(2, snap.DurationSecs);
        Assert.Null(active.Snapshot.EndedAt); // active session must not be adopted
    }

    [Fact]
    public void CorruptManifestIsSkippedNotFatal()
    {
        _store.CreateSession(localId: "good", startedAt: DateTimeOffset.UtcNow, sourceApp: null);
        string badDir = Path.Combine(_rootDir, "session-corrupt");
        Directory.CreateDirectory(badDir);
        File.WriteAllText(Path.Combine(badDir, "manifest.json"), "not json");

        var pending = _store.PendingSessions();
        Assert.Equal(new[] { "good" }, pending.Select(s => s.LocalId));
    }
}
